File: LeetCode/Container With Most Water.py
class Solution:
    # @return an integer
    def maxArea(self, height):
        l=0
        r=len(height) - 1
        maxArea=0

        while (l < r):
            # print (l,r)
            maxArea = max(maxArea, min(height[l],height[r]) * (r-l))
            if height[l] > height[r]:
                k = r - 1
                while k > l and height[k]<=height[r]:
                    k-=1
                r=k
            else:
                k = l + 1
                while k < r and height[k]<=height[l]:
                    k+=1
                l=k
        return maxArea

if __name__ == '__main__':
    s = Solution()
    print s.maxArea([4,6,2,6,7,11,2])
    print s.maxArea([])

File: LeetCode/Container With Most Water.cpp
class Solution {
public:
    int maxArea(vector<int> &height) {
        int n = height.size();
        int left = 0, right = n - 1, k;
        int maxWater = 0;
        while (left < right){
            maxWater = max(maxWater, (right - left) * min(height[right], height[left]));
            if (height[left] < height[right]){
                k = left;
                while (height[k] <= height[left]) k++;
                left = k;
            }else{
                k = right;
                while (height[k] <= height[right]) k--;
                right = k;
            }
        }
        return maxWater;
    }
};